### DISCUSSION TITLE
Mozc does not seem to work under VanillaOS

### DISCUSSION THREAD

**Discussion by 0chroma** [google/mozc#1066] [Q&A]
## Description
When using the linux distribution VanillaOS, mozc doesn't seem to work properly.

## Steps to reproduce
1. install mozc via `abroot pkg add ibus-mozc; abroot pkg apply`
2. Enable mozc in gnome configuration panel
3. Attempt to type

## Expected behavior
Mozc starts correctly and allows input

## Actual behavior
The following dialogs appear:

"Failed to get current config values."
"Cannot start conversion engine. Please restart your computer."

## Screenshots
n/a

## Version or commit-id
2.28.4715.102

## Environment
 - OS:  VanillaOS Orchid
 - IMF: ibus
 - Related Applications: all

## Investigations
 - Whether this issue happens on both Ibus and other IMF (e.g. Fcitx, uim).
   + no
 - Whether this issue happens on other IMEs (e.g. Anthy, SKK).
   + does not appear to happen in Anthy
 - What applications this issue happens on (e.g. Chromium, gedit).
   + all of them
 - What applications this issue does not happen on (e.g. Chromium, gedit).
   + none found
 - (optional) What versions or commit-ids this issue did not happen on
   (e.g. Mozc-2.28.4960.100+24.11.oss).
   + 2.28.4715.102

## Additional context
It appears that it may be related to some sort of IPC setup, see the following strace excerpt:
```
......
[pid  6566] 01:25:17 openat(AT_FDCWD, "/etc/passwd", O_RDONLY|O_CLOEXEC) = 6
[pid  6566] 01:25:17 newfstatat(6, "", {st_mode=S_IFREG|0644, st_size=2670, ...}, AT_EMPTY_PATH) = 0
[pid  6566] 01:25:17 openat(AT_FDCWD, "/dev/shm/sem.fc32c9369a38", O_RDWR|O_NOFOLLOW|O_CLOEXEC) = -1 ENOENT (No such file or directory)
strace: Process 6569 attached
[pid  6566] 01:25:17 openat(AT_FDCWD, "/proc/self/task/6569/comm", O_RDWR) = 6
[libprotobuf ERROR google/protobuf/message_lite.cc:133] Can't parse message of type "mozc.commands.Input" because it is missing required fields: type
[libprotobuf ERROR google/protobuf/message_lite.cc:133] Can't parse message of type "mozc.commands.Input" because it is missing required fields: type
......
```

From what I know about VanillaOS, there's two major differences that could cause this. Either mozc doesn't like that `/usr` is a link to `/.system/usr`, or there's some file in the root FS it's expecting read/write access to but can only get read access.

The relevant bug report in VanillaOS is here, many of us would be happy to help diagnose further! https://github.com/Vanilla-OS/live-iso/issues/68

**Comment by hiroyuki-komatsu** [comment]
Hi 0chroma,
Thank you for the report.

* 2.28.4715.102 is an old version more than 2 years ago.
* If protobuf is linked as a shared library, we highly recommend to link protobuf as a static library.

Could you please try the recent version with statically linked protobuf?

FYI. the log files are available at ~/.config/mozc/ (or ~/.mozc)

**Comment by 0chroma** [comment]
thank you for the hint! I'll take a look at how the protobuf libs are being linked when I can

**Comment by yukawa** [comment]
Was there any update on this?  I'm converting this to discussion as it seems there is no immediate action item in the Mozc side.

**Comment by 0chroma** [comment]
no updates yet, since it's a debian package I don't have a ton of control over how the libraries are linked. It does work on debian though, I'm wondering if maybe a socket isn't getting shared correctly or something

**Comment by 0chroma** [comment]
I figured out a workaround for anyone looking for a solution:

- install mozc-ibus as usual `abroot pkg add ibus-mozc && abroot pkg apply`
- make a vanillaos subsystem in apx, install ibus-mozc in there `sudo apt install ibus-mozc`
- add a `mozc_server.desktop` file to `~/.config/autostart`
```
[Desktop Entry]
Type=Application
Name=mozc_server
Exec=/usr/bin/apx mozc run mozc_server
```

should work after adding mozc through gnome. I still haven't figured out why I'm getting protobuf errors when running it on the host sadly.
